### ISSUE TITLE
WorkspaceEdit is incompatible with the protocol

### ISSUE THREAD

**Issue by KalitaAlexey** [Microsoft/vscode-languageserver-node#173]
https://github.com/Microsoft/vscode-languageserver-node/blob/master/types/src/main.ts#L366
```ts
export interface WorkspaceEdit {
	changes: TextDocumentEdit[];
}
```
In the protocol it described as follows:
```ts
interface WorkspaceEdit {
    /**
     * Holds changes to existing resources.
     */
    changes: { [uri: DocumentUri]: TextEdit[]; };
}
```

Also [asWorkspaceEdit](https://github.com/Microsoft/vscode-languageserver-node/blob/master/client/src/protocolConverter.ts#L418) is wrong.

This is cause of the problem in my extension.

It worked on `"vscode-languageclient": "^2.6.3"`, but it doesn't on `"vscode-languageclient": "3.0.4"`.

**Comment by gorkem** [comment]
@dbaeumer https://github.com/Microsoft/vscode-languageserver-node/commit/75c63564a3a678d8fc39dd63117b8e0be645cbe7 seems to have broken the [asWorkspaceEdit](https://github.com/Microsoft/vscode-languageserver-node/blob/master/client/src/protocolConverter.ts#L418). This is now a blocker for https://github.com/redhat-developer/vscode-java/issues/99 since we can not move to the 3.x version of the library.

**Comment by dbaeumer** [comment]
@gorkem I will have a look. I know that we discussed this but I need to refresh my memory.

**Comment by dbaeumer** [comment]
Reading through the code I remember why we changed this for the 3.0 version of the protocol. The reason was that the old `WorkspaceEdit` was a map from URI to edit. The problem was that this didn't allow to specify on which version of a document an edit is applicable. This is why we changed this from a map to an array where the `TextDocumentEdit` contains a `textDocument` property pointing to a `VersionedTextDocumentIdentifier`.

I know this is breaking but to support proper version checking when applying text edits this was the best way to go forward. Now users of the protocol have to think about this.

So bottom line is that the spec is incorrect. I will fix this. Would you be able to adopt this?

**Comment by gorkem** [comment]
I am concerned that this will break all the existing implementations, not all clients are able to move as quick.  I think we need a migration path on this. Perhaps we can leave the changes as a deprecated field and use something like `edits` for the `TextDocumentEdit`. 

I think we should consider https://github.com/Microsoft/language-server-protocol/issues/41 before making changes to `WorkspaceEdit`on the protocol.

The chaps on LSP4J @svenefftinge or @akosyakov can tell how quickly they can respond to such changes. It is pretty quick for jdt.ls once LSP4J has a build with adopting changes

**Comment by dbaeumer** [comment]
I updated the documentation / spec of the protocol.

**Comment by dbaeumer** [comment]
The old clients is a fair argument. Addressing this should be done like all the new features in 3.0.x. We should have a client capability property specifying which type of work space edit the client expects together. If the new format is not supported by the client then we should return the old structure. I missed this for this change but did it for all new features for 3.0.x. Don't know why I forgot it here. Will look into this this week.

**Comment by dbaeumer** [comment]
What I will do is the following. I will declare WorkspaceEdit as follows:

```ts
/**
 * A workspace edit represents changes to many resources managed
 * in the workspace. The edit should either provide `changes` or
 * `documentChanges`. If documentChanges are present they are preferred
 * over `changes`.
 */
export interface WorkspaceEdit {
	/**
	 * Holds changes to existing resources.
	 */
	changes?: { [uri: string]: TextEdit[]; };

	/**
	 * An array of `TextDocumentEdit`s to express changes
	 * to specific a specific version of a text document.
	 */
	documentChanges?: TextDocumentEdit[];
}```

and add a ClientCapability:

```ts
	versionedWorkspaceEdit?: boolean;
```

If set the by the client the server is allowed to send TextDocumentEdit arrays in the property documentChanges.

**Comment by dbaeumer** [comment]
The correct client capability will look like:
```ts
export interface WorkspaceClientCapabilites {
        ...
	/**
	 * Capabilities specific to `WorkspaceEdit`s
	 */
	workspaceEdit?: {
		/**
		 * The client supports versioned document changes in `WorkspaceEdit`s
		 */
		documentChanges?: boolean;
	};
        ...
}
```

**Comment by dbaeumer** [comment]
A fix is available on release/3.2.0-alpha.1. Need to review tomorrow and then deliver a next release.

**Comment by akosyakov** [comment]
@gorkem lsp4j is updated according to change in the protocol: https://github.com/eclipse/lsp4j/commit/46b8d7a3e57bd70a3edc33e7483da1b8e5112e43

**Comment by dbaeumer** [comment]
I pushed a 3.2.0-alpha.x version to npm under the next tag in case you want to give it a try.
